File: packages/python-sdk/vaultkey/wallets.py
"""Wallets resource — create, retrieve, sign, balance, broadcast, sweep."""
from __future__ import annotations

from typing import Dict, Optional, Tuple

from .signing import Signing
from .types import (
    APIError,
    BroadcastEVMResult,
    BroadcastSolanaResult,
    EVMBalance,
    SigningJob,
    SolanaBalance,
    SweepParams,
    Wallet,
    WalletCreate,
    WalletList,
)


class Wallets:
    """Wallet management and operations.

    Example
    -------
    ```python
    vk = VaultKey(api_key="vk_live_...", api_secret="...")

    # Create
    wallet, err = vk.wallets.create({"user_id": "user_123", "chain_type": "evm"})

    # Get
    wallet, err = vk.wallets.get("wallet_id")

    # List by user
    result, err = vk.wallets.list_by_user("user_123")

    # Sign
    job, err = vk.wallets.signing.evm_message("wallet_id", {"payload": {...}})

    # Balance
    bal, err = vk.wallets.evm_balance("wallet_id", chain_name="base")

    # Broadcast
    result, err = vk.wallets.broadcast_evm("wallet_id", "0x...", chain_name="base")

    # Sweep
    job, err = vk.wallets.sweep("wallet_id", {"chain_type": "evm", "chain_name": "base"})
    ```
    """

    def __init__(self, client: "VaultKey") -> None:
        self._client = client
        self.signing = Signing(client)

    def create(
        self, params: WalletCreate
    ) -> Tuple[Optional[Wallet], Optional[APIError]]:
        """Create a new wallet for a user.

        Parameters
        ----------
        params:
            Dict with ``user_id`` (required), ``chain_type`` (required,
            ``"evm"`` or ``"solana"``), and optional ``label``.
        """
        body: Dict = {
            "user_id": params["user_id"],
            "chain_type": params["chain_type"],
        }
        if params.get("label"):
            body["label"] = params["label"]

        data, err = self._client.post("/wallets", body)
        return data, err  # type: ignore[return-value]

    def get(self, wallet_id: str) -> Tuple[Optional[Wallet], Optional[APIError]]:
        """Retrieve a wallet by its ID."""
        data, err = self._client.get(f"/wallets/{wallet_id}")
        return data, err  # type: ignore[return-value]

    def list_by_user(
        self,
        user_id: str,
        *,
        after: Optional[str] = None,
        limit: Optional[int] = None,
    ) -> Tuple[Optional[WalletList], Optional[APIError]]:
        """List all wallets for a user. Results are paginated.

        Parameters
        ----------
        user_id:
            The user whose wallets to list.
        after:
            Cursor from the previous page's ``next_cursor`` field.
        limit:
            Number of results per page.

        Example
        -------
        ```python
        result, err = vk.wallets.list_by_user("user_123")
        if result["has_more"]:
            page2, err = vk.wallets.list_by_user(
                "user_123", after=result["next_cursor"]
            )
        ```
        """
        params: Dict[str, str] = {}
        if after:
            params["after"] = after
        if limit is not None:
            params["limit"] = str(limit)

        data, err = self._client.get(f"/users/{user_id}/wallets", params=params or None)
        return data, err  # type: ignore[return-value]

    def evm_balance(
        self,
        wallet_id: str,
        *,
        chain_name: Optional[str] = None,
        chain_id: Optional[str] = None,
    ) -> Tuple[Optional[EVMBalance], Optional[APIError]]:
        """Get the native token balance for an EVM wallet.

        Provide ``chain_name`` (preferred) or ``chain_id``.
        ``chain_name`` takes precedence if both are supplied.

        Example
        -------
        ```python
        bal, err = vk.wallets.evm_balance("wallet_id", chain_name="base")
        print(bal["balance"])  # "0.05"
        ```
        """
        params: Dict[str, str] = {}
        if chain_name:
            params["chain_name"] = chain_name
        elif chain_id:
            params["chain_id"] = chain_id

        data, err = self._client.get(
            f"/wallets/{wallet_id}/balance/evm", params=params or None
        )
        return data, err  # type: ignore[return-value]

    def solana_balance(
        self, wallet_id: str
    ) -> Tuple[Optional[EVMBalance], Optional[APIError]]:
        """Get the SOL balance for a Solana wallet.

        Example
        -------
        ```python
        bal, err = vk.wallets.solana_balance("wallet_id")
        print(bal["balance"])  # "1.5"
        ```
        """
        data, err = self._client.get(f"/wallets/{wallet_id}/balance/solana")
        return data, err  # type: ignore[return-value]

    def broadcast_evm(
        self,
        wallet_id: str,
        signed_tx: str,
        *,
        chain_name: Optional[str] = None,
        chain_id: Optional[str] = None,
    ) -> Tuple[Optional[BroadcastEVMResult], Optional[APIError]]:
        """Broadcast a pre-signed EVM transaction.

        Provide ``chain_name`` (preferred) or ``chain_id``.

        Example
        -------
        ```python
        result, err = vk.wallets.broadcast_evm(
            "wallet_id", "0x...", chain_name="base"
        )
        print(result["tx_hash"])
        ```
        """
        body: Dict = {"signed_tx": signed_tx}
        if chain_name:
            body["chain_name"] = chain_name
        elif chain_id:
            body["chain_id"] = chain_id

        data, err = self._client.post(f"/wallets/{wallet_id}/broadcast", body)
        return data, err  # type: ignore[return-value]

    def broadcast_solana(
        self,
        wallet_id: str,
        signed_tx: str,
    ) -> Tuple[Optional[BroadcastSolanaResult], Optional[APIError]]:
        """Broadcast a pre-signed Solana transaction.

        Example
        -------
        ```python
        result, err = vk.wallets.broadcast_solana("wallet_id", "base58tx...")
        print(result["signature"])
        ```
        """
        data, err = self._client.post(
            f"/wallets/{wallet_id}/broadcast", {"signed_tx": signed_tx}
        )
        return data, err  # type: ignore[return-value]

    def sweep(
        self,
        wallet_id: str,
        params: SweepParams,
    ) -> Tuple[Optional[SigningJob], Optional[APIError]]:
        """Trigger a sweep — move all funds to the configured master wallet.

        The operation is async. Poll ``vk.jobs.get(job_id)`` for the result.

        Parameters
        ----------
        wallet_id:
            The wallet to sweep from.
        params:
            Dict with ``chain_type`` (required). For EVM, provide
            ``chain_name`` (preferred) or ``chain_id``.

        Example
        -------
        ```python
        # EVM sweep
        job, err = vk.wallets.sweep(
            "wallet_id",
            {"chain_type": "evm", "chain_name": "base"},
        )

        # Solana sweep
        job, err = vk.wallets.sweep("wallet_id", {"chain_type": "solana"})
        ```
        """
        body: Dict = {"chain_type": params["chain_type"]}

        if params["chain_type"] == "evm":
            # chain_name takes precedence — mirrors Go resolveEVMChain logic
            if params.get("chain_name"):
                body["chain_name"] = params["chain_name"]
            elif params.get("chain_id"):
                body["chain_id"] = params["chain_id"]

        data, err = self._client.post(f"/wallets/{wallet_id}/sweep", body)
        return data, err  # type: ignore[return-value]


from .vaultkey import VaultKey  # noqa: E402

File: packages/sdk/src/wallets.ts
import { VaultKey } from "./vaultkey";
import { Signing } from "./signing";
import {
  VaultKeyResponse,
  Wallet,
  CreateWalletParams,
  ListWalletsResult,
  PaginationParams,
  EVMBalanceResult,
  SolanaBalanceResult,
  BroadcastEVMResult,
  BroadcastSolanaResult,
  SweepParams,
  SigningJob,
  EVMChainParams,
} from "../types";

// ── Internal request shapes (snake_case to match Go API) ─────────────────────

type CreateWalletBody = {
  user_id: string;
  chain_type: string;
  label?: string;
};

type BroadcastBody = {
  signed_tx: string;
  chain_name?: string;
  chain_id?: string;
};

/**
 * Wallet management — create wallets, retrieve them, sign messages,
 * check balances, broadcast transactions, and trigger sweeps.
 *
 * @example
 * const vaultkey = new VaultKey({ apiKey: "vk_live_...", apiSecret: "..." });
 *
 * // Create a wallet
 * const { data } = await vaultkey.wallets.create({
 *   userId: "user_123",
 *   chainType: "evm",
 * });
 *
 * // Sign a message
 * const { data: job } = await vaultkey.wallets
 *   .signing("wallet_id")
 *   .evmMessage({ payload: { message: "Hello" } });
 *
 * // Poll until done
 * const { data: result } = await vaultkey.jobs.get(job.jobId);
 */
export class Wallets {
  constructor(private readonly client: VaultKey) {}

  /**
   * Create a new wallet for a user.
   *
   * @example
   * const { data, error } = await vaultkey.wallets.create({
   *   userId: "user_123",
   *   chainType: "evm",
   *   label: "Primary wallet",
   * });
   */
  async create(params: CreateWalletParams): Promise<VaultKeyResponse<Wallet>> {
    const body: CreateWalletBody = {
      user_id: params.userId,
      chain_type: params.chainType,
    };
    if (params.label) body.label = params.label;

    return this.client.post<Wallet>("/wallets", body);
  }

  /**
   * Retrieve a wallet by its ID.
   *
   * @example
   * const { data } = await vaultkey.wallets.get("wallet_abc123");
   */
  async get(walletId: string): Promise<VaultKeyResponse<Wallet>> {
    return this.client.get<Wallet>(`/wallets/${walletId}`);
  }

  /**
   * List all wallets belonging to a user.
   * Results are paginated — use `nextCursor` to fetch the next page.
   *
   * @example
   * const { data } = await vaultkey.wallets.listByUser("user_123");
   * // Next page:
   * const { data: page2 } = await vaultkey.wallets.listByUser("user_123", {
   *   after: data.nextCursor,
   * });
   */
  async listByUser(
    userId: string,
    pagination?: PaginationParams
  ): Promise<VaultKeyResponse<ListWalletsResult>> {
    const params = new URLSearchParams();
    if (pagination?.after) params.set("after", pagination.after);
    if (pagination?.limit) params.set("limit", String(pagination.limit));
    const qs = params.toString();
    return this.client.get<ListWalletsResult>(
      `/users/${userId}/wallets${qs ? `?${qs}` : ""}`
    );
  }

  /**
   * Returns a signing interface scoped to the given wallet.
   * Use this to sign EVM or Solana messages.
   *
   * @example
   * const { data } = await vaultkey.wallets
   *   .signing("wallet_id")
   *   .evmMessage({ payload: { message: "Hello" } });
   */
  signing(walletId: string): Signing {
    return new Signing(this.client, walletId);
  }

  /**
   * Get the native token balance for an EVM wallet.
   * Provide chainName (preferred) or chainId.
   *
   * @example
   * const { data } = await vaultkey.wallets.evmBalance("wallet_id", {
   *   chainName: "base",
   * });
   * console.log(data.balance); // "0.05"
   */
  async evmBalance(
    walletId: string,
    chain: EVMChainParams
  ): Promise<VaultKeyResponse<EVMBalanceResult>> {
    const params = new URLSearchParams();
    if ("chainName" in chain && chain.chainName) {
      params.set("chain_name", chain.chainName);
    } else if ("chainId" in chain && chain.chainId) {
      params.set("chain_id", chain.chainId);
    }
    return this.client.get<EVMBalanceResult>(
      `/wallets/${walletId}/balance/evm?${params.toString()}`
    );
  }

  /**
   * Get the SOL balance for a Solana wallet.
   *
   * @example
   * const { data } = await vaultkey.wallets.solanaBalance("wallet_id");
   * console.log(data.balance); // "1.5"
   */
  async solanaBalance(
    walletId: string
  ): Promise<VaultKeyResponse<SolanaBalanceResult>> {
    return this.client.get<SolanaBalanceResult>(
      `/wallets/${walletId}/balance/solana`
    );
  }

  /**
   * Broadcast a pre-signed EVM transaction.
   * Provide chainName (preferred) or chainId.
   *
   * @example
   * const { data } = await vaultkey.wallets.broadcastEVM("wallet_id", {
   *   signedTx: "0x...",
   *   chainName: "base",
   * });
   * console.log(data.txHash);
   */
  async broadcastEVM(
    walletId: string,
    params: { signedTx: string } & EVMChainParams
  ): Promise<VaultKeyResponse<BroadcastEVMResult>> {
    const body: BroadcastBody = { signed_tx: params.signedTx };
    if ("chainName" in params && params.chainName) {
      body.chain_name = params.chainName;
    } else if ("chainId" in params && params.chainId) {
      body.chain_id = params.chainId;
    }
    return this.client.post<BroadcastEVMResult>(
      `/wallets/${walletId}/broadcast`,
      body
    );
  }

  /**
   * Broadcast a pre-signed Solana transaction.
   *
   * @example
   * const { data } = await vaultkey.wallets.broadcastSolana("wallet_id", {
   *   signedTx: "base58encodedtx...",
   * });
   * console.log(data.signature);
   */
  async broadcastSolana(
    walletId: string,
    params: { signedTx: string }
  ): Promise<VaultKeyResponse<BroadcastSolanaResult>> {
    return this.client.post<BroadcastSolanaResult>(
      `/wallets/${walletId}/broadcast`,
      { signed_tx: params.signedTx }
    );
  }

  /**
   * Trigger a sweep — move all funds from this wallet to the configured
   * master wallet. The operation is asynchronous; poll via `vaultkey.jobs.get`.
   *
   * @example
   * // EVM sweep
   * const { data } = await vaultkey.wallets.sweep("wallet_id", {
   *   chainType: "evm",
   *   chainName: "base",
   * });
   *
   * // Solana sweep
   * const { data } = await vaultkey.wallets.sweep("wallet_id", {
   *   chainType: "solana",
   * });
   */
  async sweep(
    walletId: string,
    params: SweepParams
  ): Promise<VaultKeyResponse<SigningJob>> {
    const body: Record<string, string> = {
      chain_type: params.chainType,
    };

    if (params.chainType === "evm") {
      // chainName takes precedence — mirrors Go resolveEVMChain logic
      if (params.chainName) {
        body.chain_name = params.chainName;
      } else if (params.chainId) {
        body.chain_id = params.chainId;
      }
    }

    return this.client.post<SigningJob>(`/wallets/${walletId}/sweep`, body);
  }
}